src/components/interface/cbuf_mgr/cos_alloc.c

```diff
@@ -0,0 +1,536 @@
+void* __libc_calloc(size_t nmemb, size_t _size) {
+  register size_t size=_size*nmemb;
+  if (nmemb && size/nmemb!=_size) {
+    (*__errno_location())=ENOMEM;
+    return 0;
+  }
+  return malloc(size);
+}
+void* calloc(size_t nmemb, size_t _size) __attribute__((weak,alias("__libc_calloc")));
+
+void* __libc_realloc(void* ptr, size_t _size);
+void* __libc_realloc(void* ptr, size_t _size) {
+  register size_t size=_size;
+  if (ptr) {
+    if (size) {
+      __alloc_t* tmp=BLOCK_START(ptr);
+      size+=sizeof(__alloc_t);
+      if (size<sizeof(__alloc_t)) goto retzero;
+      size=(size<=__MAX_SMALL_SIZE)?GET_SIZE(size):PAGE_ALIGN(size);
+      if (tmp->size!=size) {
+	if ((tmp->size<=__MAX_SMALL_SIZE)) {
+	  void *new=_alloc_libc_malloc(_size);
+	  if (new) {
+	    register __alloc_t* foo=BLOCK_START(new);
+	    size=foo->size;
+	    if (size>tmp->size) size=tmp->size;
+	    if (size) memcpy(new,ptr,size-sizeof(__alloc_t));
+	    _alloc_libc_free(ptr);
+	  }
+	  ptr=new;
+	}
+	else {
+	  register __alloc_t* foo;
+	  size=PAGE_ALIGN(size);
+	  foo=mremap(tmp,tmp->size,size,MREMAP_MAYMOVE);
+	  if (foo==MAP_FAILED) {
+retzero:
+	    (*__errno_location())=ENOMEM;
+	    ptr=0;
+	  }
+	  else {
+	    foo->size=size;
+	    ptr=BLOCK_RET(foo);
+	  }
+	}
+      }
+    }
+    else { /* size==0 */
+      _alloc_libc_free(ptr);
+      ptr = NULL;
+    }
+  }
+  else { /* ptr==0 */
+    if (size) {
+      ptr=_alloc_libc_malloc(size);
+    }
+  }
+  return ptr;
+}
+void* realloc(void* ptr, size_t size) __attribute__((weak,alias("__libc_realloc")));
+
+#endif
+
+/********************* testing code on unix ***********************/
+
+#ifdef UNIX_TEST
+
+#define ITER 100000
+#define PTRS_LIVE 100
+#define SIZE_LB 4
+#define SIZE_UB 8192
+#define SIZE_GET ((rand() % (SIZE_UB-SIZE_LB)) + SIZE_LB)
+
+#include <stdlib.h>
+#include <time.h>
+#include <stdio.h>
+#include <string.h>
+
+int find_ptr(long *ptrs, int sz, int empty)
+{
+	int i; 
+
+	for (i = 0 ; i < sz ; i++) {
+		if (empty && ptrs[i] == 0) return i;
+		else if (!empty && ptrs[i] != 0) return i;
+	}
+
+	return -1;
+}
+
+void malloc_rand(long *ptrs, int sz)
+{
+	int idx = find_ptr(ptrs, sz, 1);
+	long *ptr;
+	int asz;
+
+	if (idx < 0) printf("Hmm, index errors when mallocing.\n");
+
+	ptr = &ptrs[idx];
+
+	asz = SIZE_GET;
+	*ptr = (long)malloc(asz);
+	if (*ptr == 0) {
+		printf("could not malloc!\n");
+	}
+
+	*(int*)*ptr = asz;
+	memset((char*)(*ptr)+sizeof(int), rand() % 256, asz-sizeof(int));
+	printf("+");
+	
+	return;
+}
+
+void free_rand(long *ptrs, int sz)
+{
+	int i, idx = find_ptr(ptrs, sz, 0), asz;
+	long *ptr;
+	char c, *arr;
+
+	if (idx < 0) printf("Hmm, index errors when freeing.\n");
+
+	ptr = &ptrs[idx];
+	asz = *(int*)(*ptr);
+	arr = ((char*)(*ptr)) + sizeof(int);
+	c = *arr;
+
+	for (i = 0 ; i < asz-sizeof(int) ; i++) {
+		if (((char *)arr)[i] != c) {
+			printf("we have a problem.\n");
+			break;
+		}
+	}
+
+	free((void*)*ptr);
+	*ptr = 0;
+	printf("-");
+
+	return;
+}
+
+int main(void)
```

Comment:
You might consider testing malloc with something that looks like this, though.  Seems more complete than your tests.  `rand()` will need to be emulated with `rdtscll`.